File: codegen/features/xsls/backendless-codegen-xsls/ios/backendless-codegen/ObjC/CRUD/CRUD/ViewControllers/RetrieveViewController.m

#import "RetrieveViewController.h"
#import "CodeSampleViewController.h"
#import "SortViewController.h"
#import "RelatedViewController.h"
#import "Utils.h"

#define BASIC_FIND @"Basic find"
#define FIND_FIRST @"Find first"
#define FIND_LAST @"Find last"
#define FIND_SORT @"Find with sort"
#define FIND_RELATED @"Find with related"

#define SHOW_SAMPLE @"ShowCodeSampleFromRetrieve"
#define SHOW_SORT @"ShowSortBy"
#define SHOW_RELATED @"ShowRelated"

@interface RetrieveViewController() {
    NSArray<NSString *> *retrieveTypes;
    NSArray<NSString *> *sortBy;
}
@end

@implementation RetrieveViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    retrieveTypes = @[BASIC_FIND, FIND_FIRST, FIND_LAST, FIND_SORT, FIND_RELATED];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return retrieveTypes.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"RetrieveTypeCell" forIndexPath:indexPath];
    cell.textLabel.text = [retrieveTypes objectAtIndex:indexPath.row];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if ([[retrieveTypes objectAtIndex:indexPath.row] isEqualToString:FIND_SORT]) {
        [self performSegueWithIdentifier:SHOW_SORT sender:nil];
    }
    else if ([[retrieveTypes objectAtIndex:indexPath.row] isEqualToString:FIND_RELATED]) {
        [self performSegueWithIdentifier:SHOW_RELATED sender:nil];
    }
    else {
        [self performSegueWithIdentifier:SHOW_SAMPLE sender:indexPath];
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([[segue identifier] isEqualToString:SHOW_SORT]) {
        SortViewController *sortVC = [segue destinationViewController];
        sortVC.tableName = self.tableName;
        [Backendless.shared.data describeWithTableName:self.tableName responseHandler:^(NSArray *properties) {
            NSMutableArray<NSString *> *fields = [NSMutableArray<NSString *> new];
            for (ObjectProperty *property in properties) {
                if (![[property getTypeName] isEqualToString:@"RELATION"] &&
                    ![[property getTypeName] isEqualToString:@"RELATION_LIST"]) {
                    [fields addObject:property.name];
                }
            }
            sortVC.fields = fields;
            [sortVC.tableView reloadData];
        } errorHandler:^(Fault *fault) {
            [utils showErrorAlert:self fault:fault];
        }];
    }
    else if ([[segue identifier] isEqualToString:SHOW_RELATED]) {
        RelatedViewController *relatedVC = [segue destinationViewController];
        relatedVC.tableName = self.tableName;
        [Backendless.shared.data describeWithTableName:self.tableName responseHandler:^(NSArray *properties) {
            NSMutableArray<NSString *> *fields = [NSMutableArray<NSString *> new];
            for (ObjectProperty *property in properties) {
                if ([[property getTypeName] isEqualToString:@"RELATION"] ||
                    [[property getTypeName] isEqualToString:@"RELATION_LIST"]) {
                    [fields addObject:property.name];
                }
            }
            relatedVC.fields = fields;
            [relatedVC.tableView reloadData];
        } errorHandler:^(Fault *fault) {
            [utils showErrorAlert:self fault:fault];
        }];
    }
    else if ([[segue identifier] isEqualToString:SHOW_SAMPLE]) {
        CodeSampleViewController *codeSampleVC = [segue destinationViewController];
        codeSampleVC.tableName = self.tableName;
        codeSampleVC.operation = [retrieveTypes objectAtIndex:((NSIndexPath *)sender).row];
    }
}

@end

File: codegen/features/xsls/backendless-codegen-xsls/ios/backendless-codegen/Swift/CRUD/CRUD/ViewControllers/RetrieveViewController.swift

import UIKit
import Backendless

class RetrieveViewController: UITableViewController {
    
    var tableName = ""
    
    private var retrieveTypes = [String]()
    private var sortBy = [String]()
    
    private let BASIC_FIND = "Basic find"
    private let FIND_FIRST = "Find first"
    private let FIND_LAST = "Find last"
    private let FIND_SORT = "Find with sort"
    private let FIND_RELATED = "Find with related"
    
    private let SHOW_SAMPLE = "ShowCodeSampleFromRetrieve"
    private let SHOW_SORT = "ShowSortBy"
    private let SHOW_RELATED = "ShowRelated"
    
    private let utils = Utils.shared
    
    override func viewDidLoad() {
        super.viewDidLoad()
        retrieveTypes = [BASIC_FIND, FIND_FIRST, FIND_LAST, FIND_SORT, FIND_RELATED]
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return retrieveTypes.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "RetrieveTypeCell", for: indexPath)
        cell.textLabel?.text = retrieveTypes[indexPath.row]
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if retrieveTypes[indexPath.row] == FIND_SORT {
            performSegue(withIdentifier: SHOW_SORT, sender: nil)
        }
        else if retrieveTypes[indexPath.row] == FIND_RELATED {
            performSegue(withIdentifier: SHOW_RELATED, sender: nil)
        }
        else {
            performSegue(withIdentifier: SHOW_SAMPLE, sender: indexPath)
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == SHOW_SORT,
            let sortVC = segue.destination as? SortViewController {
            sortVC.tableName = tableName
            Backendless.shared.data.describe(tableName: tableName, responseHandler: { properties in
                var fields = [String]()
                for property in properties {
                    if property.getTypeName() != "RELATION" && property.getTypeName() != "RELATION_LIST" {
                        fields.append(property.name)
                    }
                }
                sortVC.fields = fields
                sortVC.tableView.reloadData()
            }, errorHandler: { [weak self] fault in
                if self != nil {
                    self?.utils.showErrorAler(target: self!, fault: fault)
                }
            })
        }
        else if segue.identifier == SHOW_RELATED {
            if let relatedVC = segue.destination as? RelatedViewController {
                relatedVC.tableName = tableName
                Backendless.shared.data.describe(tableName: tableName, responseHandler: { properties in
                    var fields = [String]()
                    for property in properties {
                        if property.getTypeName() == "RELATION" || property.getTypeName() == "RELATION_LIST" {
                            fields.append(property.name)
                        }
                    }
                    relatedVC.fields = fields
                    relatedVC.tableView.reloadData()
                }, errorHandler: { [weak self] fault in
                    if self != nil {
                        self?.utils.showErrorAler(target: self!, fault: fault)
                    }
                })
            }
        }
        else if segue.identifier == SHOW_SAMPLE {
            if let codeSampleVC = segue.destination as? CodeSampleViewController,
                let indexPath = sender as? IndexPath {
                codeSampleVC.tableName = tableName
                codeSampleVC.operation = retrieveTypes[indexPath.row]
            }
        }
    }
}
